karapace/kafka_rest_apis/admin.py

```diff
@@ -36,53 +44,59 @@ def cluster_metadata(self, topics: List[str] = None) -> dict:
         """List all kafka topics."""
         resp = {"topics": {}}
         brokers = set()
+        node_found = True
         metadata_version = self._matching_api_version(MetadataRequest)
-        if 1 <= metadata_version <= 6:
-            if not topics:
-                request = MetadataRequest[metadata_version]()
-            else:
-                request = MetadataRequest[metadata_version](topics=topics)
-            future = self._send_request_to_node(self._client.least_loaded_node(), request)
-            self._wait_for_futures([future])
-            response = future.value
-            resp_brokers = response.brokers
-            for b in resp_brokers:
-                if metadata_version == 0:
-                    node_id, _, _ = b
-                else:
-                    node_id, _, _, _ = b
-                brokers.add(node_id)
-            resp["brokers"] = list(brokers)
-            if not response.topics:
-                return resp
+        if metadata_version > 6 or metadata_version < 1:
+            raise UnrecognizedBrokerVersion(
+                "Kafka Admin interface cannot determine the controller using MetadataRequest_v{}.".format(metadata_version)
+            )
+        request = MetadataRequest[metadata_version](topics=topics)
+        with self.client_lock:
+            node_id = self._client.least_loaded_node()
+            # normal admin code will call poll on client not ready which will sometime block indefinitely on select
+            if not self.first_boot or not self._client.ready(node_id):
+                #  pylint: disable=protected-access
+                node_id = list(self._client.cluster._bootstrap_brokers.keys())[0]
+                self.first_boot = False
+                node_found = False
+            future = self._send_request_to_node(node_id, request)
+            self.log.info("Retrieved metadata using node id %s", node_id)
+        self._wait_for_futures([future])
+        response = future.value
+        if not node_found:
+            self._client.cluster.update_metadata(response)
+        resp_brokers = response.brokers
+        for b in resp_brokers:
+            node_id, _, _, _ = b
+            brokers.add(node_id)
+        resp["brokers"] = list(brokers)
+        if not response.topics:
+            return resp
 
-            for tup in response.topics:
-                if response.API_KEY != 0:
-                    err, topic, _, partitions = tup
+        for tup in response.topics:
```

Comment:
This is a fairly long method, don't see a nice way of splitting it up but thought I'd mention it.

Reply:
refactored after some updates, into get and create method (turns out we can only update cluster metadata using V1 requests or similar)